### PR TITLE
PCHR-1427: Implement file uploader for leave instance

### PR DESCRIPTION
**Pull request by dhavalsharma**
The uploader attribute is added to defaultCustomData. It is also removed in toAPI calls of LeaveRequestInstance.

The create and update methods are modified to call the uploadAll method which will upload any files which are in FileUploader's queue.

## Related changes

1. All test files dependent on LeaveRequestInstance changed to accommodate dependency on FileUpload.
2. FileUpload mock file enhanced to return promise on onCompleteAll.
3. removeComment functionality moved to instance 

## Changes for testing 

1. Temp fix was added to uk.co.compucorp.civicrm.hrcomments/hrcomments.php 
```javascript

function hrcomments_civicrm_alterAPIPermissions($entity, $action, &$params, &$permissions) {
  $actionEntities = [
    'create' => ['comment'],
    'delete' => ['comment'],
    'get' => ['comment']
  ];
  foreach ($actionEntities as $action => $entities) {
    foreach ($entities as $entity) {
      $permissions[$entity][$action] = ['access AJAX API'];
    }
  }
}
```
2. `CiviHR: view all contacts` permission was given to staff and manager to resolve ACL issues.

The above 2 will be fixed by this [PR](https://github.com/civicrm/civihr/pull/1644) 

## Demo
A demo is attached here

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js [hunk 1]
```diff
@@ -5,6 +5,7 @@ define([
   'leave-absences/shared/modules/apis',
   'leave-absences/shared/modules/models-instances',
   'leave-absences/shared/modules/shared-settings',
+  'common/angular-file-upload'
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js]: Keep common dependencies on top, then leave and absences.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js [hunk 2]
```diff
@@ -13,6 +14,7 @@ define([
     'common.services',
     'leave-absences.apis',
     'leave-absences.models.instances',
-    'leave-absences.settings'
+    'leave-absences.settings',
+    'angularFileUpload'
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js]: Keep common dependencies on top, then leave and absences.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js [hunk 3]
```diff
@@ -10,7 +10,8 @@
     'mocks/apis/option-group-api-mock',
     'common/mocks/services/api/contact-mock',
     'leave-absences/shared/config',
-    'leave-absences/manager-leave/app'
+    'leave-absences/manager-leave/app',
+    'common/mocks/services/file-uploader-mock',
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js]: Keep  common dependencies first, please check all other files of this PR too

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js [hunk 4]
```diff
@@ -678,14 +685,62 @@ define([
...
+          promise = instance.create();
+        });
+
+        afterEach(function () {
+          $rootScope.$apply();
+        });
+
+        it('calls corresponding end point', function () {
+          promise.then(function () {
+            expect(instance.uploader.uploadAll).toHaveBeenCalledWith({ entityID: jasmine.any(String) });
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: Why jasmine.any(String), it can be tested with `instance.id`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js [hunk 5]
```diff
@@ -678,14 +685,62 @@ define([
...
+          promise = instance.update();
+        });
+
+        afterEach(function () {
+          $rootScope.$apply();
+        });
+
+        it('calls corresponding end point', function () {
+          promise.then(function () {
+            expect(instance.uploader.uploadAll).toHaveBeenCalledWith({ entityID: jasmine.any(String) });
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: Why jasmine.any(String), it can be tested with `instance.id`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js [hunk 6]
```diff
@@ -678,14 +685,62 @@ define([
...
+          });
+        });
+      });
+
+      describe('on update()', function () {
+        beforeEach(function () {
+          requestData = helper.createRandomLeaveRequest();
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: `requestData`is already created. No need to define again.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js [hunk 7]
```diff
@@ -678,14 +685,62 @@ define([
...
+          requestData.id = '12';
+          instance = LeaveRequestInstance.init(requestData, false);
+          spyOn(instance.uploader, 'uploadAll').and.callThrough();
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: lot of duplicate code,

```javascript
describe('uploading files', function () {
  var promise;

  beforeEach(function () {
    requestData = helper.createRandomLeaveRequest();
    requestData.id = '12';
    instance = LeaveRequestInstance.init(requestData, false);
    spyOn(instance.uploader, 'uploadAll').and.callThrough();
  });

  afterEach(function () {
    $rootScope.$apply();
  });

  describe('on create()', function () {
    beforeEach(function () {
      promise = instance.create();
    });
    
    it('calls corresponding end point', function () {
      promise.then(function () {
        expect(instance.uploader.uploadAll).toHaveBeenCalledWith({ entityID: instance.id });
      });
    });
  });

  describe('on update()', function () {
    beforeEach(function () {
      promise = instance.update();
    });

    it('calls corresponding end point', function () {
      promise.then(function () {
        expect(instance.uploader.uploadAll).toHaveBeenCalledWith({ entityID: instance.id });
      });
    });
  });
});
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/leave-request-model_test.js [hunk 8]
```diff
@@ -10,12 +11,14 @@ define([
...
-      $provide = _$provide_;
+    beforeEach(module('leave-absences.models', 'leave-absences.mocks', 'common.mocks',
+      function (_$provide_) {
+        $provide = _$provide_;
     }));
 
-    beforeEach(inject(function (_LeaveRequestAPIMock_) {
+    beforeEach(inject(function (_LeaveRequestAPIMock_, _FileUploaderMock_) {
       $provide.value('LeaveRequestAPI', _LeaveRequestAPIMock_);
+      $provide.value('FileUploader', _FileUploaderMock_);
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/leave-request-model_test.js]: Is it necessary to mock `FileUploader` in each file? I commented line 21 of this file and test cases ran successfully. 
Please check all files where mock has been added.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js (hunk 7 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: the instance is not shared across describe so the one in oncreate is not available in onupdate, hence need to create it again with id

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js [hunk 9]
```diff
@@ -2,19 +2,19 @@ define([
   'common/angular',
   'common/modules/models',
   'common/modules/services',
+  'common/angular-file-upload',
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js]: 'common/angular',
'common/angular-file-upload'
'common/modules/models',
'common/modules/services',

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js [hunk 10]
```diff
@@ -11,8 +12,9 @@ define([
   return angular.module('leave-absences.models', [
     'common.models',
     'common.services',
+    'angularFileUpload',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js]: If you have the `FileUpload` service injecting the `common/angular-file-upload` module, why didn't you define the `angularFileUpload` dependency in [`common/modules/services`](https://github.com/civicrm/civihr/blob/la-milestone-3/org.civicrm.reqangular/src/common/modules/services.js) instead of doing it here?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js [hunk 11]
```diff
@@ -14,15 +14,16 @@
...
   ], function (angular, moment, _, optionGroupMock, publicHolidayData, workPatternData, leaveRequestData) {
     'use strict';
 
     describe('myLeaveCalendar', function () {
       var $compile, $log, $q, $rootScope, component, controller, sharedSettings,
         $provide, OptionGroup, OptionGroupAPIMock, Calendar, CalendarInstance, LeaveRequest;
 
-      beforeEach(module('leave-absences.templates', 'leave-absences.mocks', 'my-leave', function (_$provide_) {
+      beforeEach(module('leave-absences.templates', 'leave-absences.mocks',
+      'my-leave', 'common.mocks', function (_$provide_) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js]: What do you need this `common.mocks` module for, in this case?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/leave-request-model_test.js [hunk 12]
```diff
@@ -10,8 +10,9 @@ define([
       LeaveRequestAPI,
       $rootScope;
 
-    beforeEach(module('leave-absences.models', 'leave-absences.mocks', function (_$provide_) {
-      $provide = _$provide_;
+    beforeEach(module('leave-absences.models', 'leave-absences.mocks',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/leave-request-model_test.js]: Are these changes coming from your IDE? They don't introduce anything, they just apply some code styling rule. Can you remove them, so we can reduce the noise in this PR?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js [hunk 13]
```diff
@@ -21,10 +21,13 @@
...
-      beforeEach(module('leave-absences.templates', 'my-leave', 'leave-absences.mocks', function (_$provide_) {
-        $provide = _$provide_;
+      beforeEach(module('leave-absences.templates', 'my-leave',
+      'leave-absences.mocks',
+        function (_$provide_) {
+          $provide = _$provide_;
       }));
-      beforeEach(inject(function (AbsencePeriodAPIMock, AbsenceTypeAPIMock, EntitlementAPIMock, LeaveRequestAPIMock, HR_settingsMock) {
+      beforeEach(inject(function (AbsencePeriodAPIMock, AbsenceTypeAPIMock,
+        EntitlementAPIMock, LeaveRequestAPIMock, HR_settingsMock) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: Are these changes coming from your IDE? They don't introduce anything, they just apply some code styling rule. Can you remove them, so we can reduce the noise in this PR?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 14]
```diff
@@ -92,7 +97,11 @@ define([
         defaultCustomData: function () {
           return {
             comments: [],
-            request_type: 'leave'
+            request_type: 'leave',
+            uploader: FileUpload.uploader({
+              entityTable: 'civicrm_hrleaveandabsences_leave_request',
+              crmAttachmentToken: sharedSettings.attachmentToken
+            })
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: In the ticket you wrote

> files:[] is not required because the list of files to be uploaded is maintained by the 3rd party plugin FileUploader

but I'm not sure that's true. the 3rd party plugin is responsible for uploading the files, but when you load a leave request which already has files linked to it, how do you retrieve them? and where do you store them? that's what `files` was for. What's your alternative approach?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 15]
```diff
@@ -73,12 +78,12 @@ define([
...
             })(index);
-          } else if(comment.toBeDeleted) {
+          } else if (comment.toBeDeleted) {
             promises.push(LeaveRequestAPI.deleteComment(comment.comment_id));
           }
         });
-
-        return $q.all(promises);
+        
+        return promises.length ? $q.all(promises) : $q.resolve();
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: now restored it and test cases pass, earlier they were breaking so had to put this check

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js [hunk 16]
```diff
@@ -678,14 +685,54 @@ define([
...
+        requestData = helper.createRandomLeaveRequest();
+        instance = LeaveRequestInstance.init(requestData, false);
+        spyOn(instance.uploader, 'uploadAll').and.callThrough();
+      });
+
+      afterEach(function () {
+        $rootScope.$apply();
+      });
+
+      describe('on create()', function () {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: no files gets uploaded as there is none in queue

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js [hunk 17]
```diff
@@ -1,5 +1,6 @@
 define([
   'common/angular',
+  'common/angular-file-upload',
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js]: I was getting this before `Unknown provider: FileUploaderProvider <- FileUploader <- FileUpload <- LeaveRequestInstance <- LeaveRequest`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js (hunk 10 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/modules/models.js]: missed that one

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/leave-request-model_test.js (hunk 12 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/leave-request-model_test.js]: no, had added common.mocks earlier but did not adjust the line after

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 18]
```diff
@@ -92,7 +97,12 @@ define([
         defaultCustomData: function () {
           return {
             comments: [],
-            request_type: 'leave'
+            request_type: 'leave',
+            uploader: FileUpload.uploader({
+              entityTable: 'civicrm_hrleaveandabsences_leave_request',
+              crmAttachmentToken: sharedSettings.attachmentToken
+            }),
+            files: []
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: please remember to put the one-liner properties first and then the multi-lines ones later
```js
{
  comments: [],
  files: [],
  request_type: 'leave',
  uploader: 
}

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 19]
```diff
@@ -247,9 +264,56 @@ define([
...
+              this.files = attachments;
+            }.bind(this));
+        },
+
+        /**
+        * Deletes the given attachment from local files array. It sets a flag
+        * to delete in final update call. This can be used by the client to set
+        * file for deletion.
+        */
+        deleteAttachment: function (attachmentID) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: The name of the method (`delete..`) and the documentation ("Deletes the given attachment from local files array") are not describing what is actually going on in the method itself. You are not deleting anything, you are just marking a file to be deleted later. So which one is it? Should you mark the file for deletion, or delete it from the array?

Shouldn't this be similar to how comments are deleted? Can you talk with @deb1990 about it since he wrote the comments bits?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 20]
```diff
@@ -247,9 +264,56 @@ define([
...
+          this.files = this.files || [];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: Why is this necessary? `files` should be initialized always when calling `init()` on the instance

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 21]
```diff
@@ -247,9 +264,56 @@ define([
...
+
+          this.files = this.files.map(function (file) {
+            if(file.attachment_id == attachmentID && !file.toBeDeleted) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: wrong code style

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 22]
```diff
@@ -247,9 +264,56 @@ define([
...
+              file.toBeDeleted = true;
+            }
+            return file;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: wrong code style

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 23]
```diff
@@ -247,9 +264,56 @@ define([
...
+
+        /**
+        * Deletes the given attachment from server. It iterates through local
+        * files array to find which are to be deleted and deletes them.
+        *
+        * @return {Promise}
+        */
+        deleteAttachments: function (attachmentID) {
+          var promises = [];
+          this.files = this.files || [];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: Unnecessary, see above

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 24]
```diff
@@ -247,9 +264,56 @@ define([
...
+          });
+        },
+
+        /**
+        * Deletes the given attachment from server. It iterates through local
+        * files array to find which are to be deleted and deletes them.
+        *
+        * @return {Promise}
+        */
+        deleteAttachments: function (attachmentID) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: please discuss this as well with @deb1990 , if there are similarities between files and comments

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 25]
```diff
@@ -247,9 +264,56 @@ define([
...
+          var promises = [];
+          this.files = this.files || [];
+
+          _.forEach(this.files, function (file) {
+            if(file.toBeDeleted) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: wrong code style

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js [hunk 26]
```diff
@@ -30,11 +30,13 @@
         ContactAPIMock,
         OptionGroupAPIMock;
 
-      beforeEach(module('leave-absences.templates', 'manager-leave', 'leave-absences.mocks', function (_$provide_) {
-        $provide = _$provide_;
+      beforeEach(module('leave-absences.templates', 'manager-leave',
+        'leave-absences.mocks', 'common.mocks', function (_$provide_) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js]: why do you need `common.mocks`?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/data/leave-request-data.js [hunk 27]
```diff
@@ -487,9 +487,11 @@ define([
       return mockData.deleteComment;
     },
     getAttachments: function () {
+      //LeaveRequest.getAttachments (leave_request_id = 1)
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/data/leave-request-data.js]: what's are these comments for? you don't have them in any other mocked function

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js [hunk 28]
```diff
@@ -14,15 +14,16 @@
...
+    'leave-absences/my-leave/app',
   ], function (angular, moment, _, optionGroupMock, publicHolidayData, workPatternData, leaveRequestData) {
     'use strict';
 
     describe('myLeaveCalendar', function () {
       var $compile, $log, $q, $rootScope, component, controller, sharedSettings,
         $provide, OptionGroup, OptionGroupAPIMock, Calendar, CalendarInstance, LeaveRequest;
 
-      beforeEach(module('leave-absences.templates', 'leave-absences.mocks', 'my-leave', function (_$provide_) {
+      beforeEach(module('leave-absences.templates', 'leave-absences.mocks',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js]: pointless changes, can you revert this whole file

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js [hunk 29]
```diff
@@ -21,9 +21,12 @@
       var $compile, $q, $log, $provide, $rootScope, component, controller;
       var AbsencePeriod, AbsenceType, Entitlement, LeaveRequest, LeaveRequestInstance, OptionGroup, HR_settings, dialog;
 
-      beforeEach(module('leave-absences.templates', 'my-leave', 'leave-absences.mocks', function (_$provide_) {
-        $provide = _$provide_;
+      beforeEach(module('leave-absences.templates', 'my-leave',
+      'leave-absences.mocks',
+        function (_$provide_) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: pointless changes, can you revert this whole file

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js [hunk 30]
```diff
@@ -678,14 +687,122 @@ define([
...
+      });
+
+      afterEach(function () {
+        $rootScope.$apply();
+      });
+
+      describe('getAttachments()', function () {
+        it('initializes files array', function () {
+          promise.then(function () {
+            expect(LeaveRequestInstance.files.length).toEqual(2);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: How do you know it's `2`? can you get the value more programmatically?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js [hunk 31]
```diff
@@ -678,14 +687,122 @@ define([
...
+          });
+        });
+      });
+
+      describe('deleteAttachment()', function () {
+        it('sets flag toBeDeleted', function () {
+          promise.then(function () {
+            LeaveRequestInstance.deleteAttachment(test_id);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js]: this goes into a `beforeEach` block

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 32]
```diff
@@ -247,9 +264,56 @@ define([
...
             result[key] = this[key];
           }
+        },
+
+        /**
+        * Gets file attachments associated with this leave request
+        *
+        * @return {Promise}
+        */
+        getAttachments: function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: for comments the name of the method is `loadComment` (also, that method is missing the documentation, can you please add it?), please stick to one standard

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js (hunk 20 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: this is from the comments experience in case they are not defined, defensive code

**Comment by AkA84** [comment]
@dhavalsharma approved but make sure that your work done with attachments is consistent with the work done by @deb1990 with comments

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 33]
```diff
@@ -247,9 +269,53 @@ define([
...
+            .then(function (attachments) {
+              this.files = attachments;
+            }.bind(this));
+        },
+
+        /**
+         * Sets the flag to mark file for deletion. The file is not yet deleted
+         * from the server.
+         */
+        deleteAttachment: function (attachmentID) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: This function is not necessary here, it can be handled in the `controller`calling this model as done for comment. 
https://github.com/civicrm/civihr/blob/211ae872153157e9ca6ceabea045a00c8afe2d51/uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js#L282-L282

Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: The internal details of instance like delete flag etc., need not be exposed to controller. It should simply call the method on instance itself to mark it for deletion. what say?

Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: That is also a good idea, in that case comments code also needs to be changed. So either change in this PR or create new Ticket.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 34]
```diff
@@ -247,9 +269,53 @@ define([
...
+          this.files = this.files.map(function (file) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: Can be simplified a bit
1. Using each, 
2. removing `&& !file.toBeDeleted` condition
```
_.each(this.files, (function (file) {
   if (file.attachment_id == attachmentID) {
     file.toBeDeleted = true;
   }
   return file;
});

Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: i think its similar so leaving it as it is

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 35]
```diff
@@ -167,6 +194,37 @@ define([
         },
 
         /**
+         * Sets the flag to mark file for deletion. The file is not yet deleted
+         * from the server.
+         */
+        deleteAttachment: function (attachmentID) {
+          this.files = this.files.map(function (file) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: The logic is still different between `comments` and `attachments`. We should definitely have similar same approach for both.

I prefer comments approach, because there we dont to need to search through an array to fetch the object, instead the object is passed through parameter.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 36]
```diff
@@ -159,6 +185,7 @@ define([
           return LeaveRequestAPI.create(this.toAPI())
             .then(function (result) {
               this.id = result.id;
+              debugger;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: remove

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 37]
```diff
@@ -167,6 +194,37 @@ define([
...
+         * @param {Object} commentObj - comment object
+         */
+        removeComment: function (commentObj) {
+          //If its an already saved comment, mark a toBeDeleted flag
+          if (commentObj.comment_id) {
+            commentObj.toBeDeleted = true;
+            return;
+          }
+
+          this.comments = _.reject(this.comments, function (comment) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: Another difference in approach between comment and attachments,
For comments
1. Object is removed from array if it does not have an ID

For attachements
1. Object is only marked as deleted but not removed from array. 

Please keep similar approach in both. 
Comments is already tested, so keep that?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js (hunk 34 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: Point no 1 is more optimal as we are not re assigning the variable.
Also `map` should only be used when we need a new copy of the array(this is more cpu extensive), but here we just need to change value of a flag so `each`should be enough and more optimal.
